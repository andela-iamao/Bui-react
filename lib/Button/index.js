'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Button Component
 */
function Button(_ref) {
    var children = _ref.children,
        props = _objectWithoutProperties(_ref, ['children']);

    return _react2.default.createElement(
        'button',
        props,
        children
    );
}

Button.propTypes = {
    /**
     * Child elements for the button component
     */
    children: _propTypes2.default.node.isRequired,
    onClick: _propTypes2.default.func
};

exports.default = Button;