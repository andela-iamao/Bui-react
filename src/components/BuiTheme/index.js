import React from 'react';
import PropTypes from 'prop-types';


class BuiTheme extends React.Component {

    getChildContext() {
        let yaml = {}
        if (!this.props.theme) {
            yaml = require('../assets/themes/default.theme.yml').json;
        } else {
            yaml = this.props.theme;
        }
        const { name, colors, ...buiThemes } = yaml;
        return {
            bui:
                {
                    theme: buiThemes
                }
        };
    }

    render() {
        return this.props.children;
    }
}

BuiTheme.childContextTypes = {
    bui: PropTypes.shape({
        theme: PropTypes.shape({
            button: PropTypes.object
        }) 
    }),
    colors: PropTypes.shape({
        primary: PropTypes.string,
        secondary: PropTypes.string
    }),
    button: PropTypes.shape({
        base: PropTypes.object
    })
};

BuiTheme.propTypes = {
    theme: PropTypes.shape({
        name: PropTypes.string,
        colors: PropTypes.shape({
            primary: PropTypes.string,
            secondary: PropTypes.string
        }),
        button: PropTypes.object
    }),
    children: PropTypes.node
};

export default BuiTheme;
