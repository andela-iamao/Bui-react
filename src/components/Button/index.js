import React from 'react';
import Radium from 'radium';
import PropTypes from 'prop-types';
import Icon from 'bui-react/Icon';

/**
 * Button Component
 */
function Button({
  icon,
  link,
  type,
  text,
  style,
  rounded,
  children,
  iconStyle,
  iconPosition,
  iconWrapperStyle,
  ...props
}) {
    const combineProps = {
      ...props,
      style:[
        styles.base,
        styles[type],
        rounded ?
          styles['rounded'](typeof rounded === 'boolean' ? null : rounded)
          :
          styles['block'],
        style
      ]
    };

    const child = (
      <span>
        {text ? text : children}
        {
          icon
            &&
            iconPosition === 'right'
              &&
              <span style={[styles.iconContainer, iconWrapperStyle ]}>
                <Icon size={20} icon={icon} style={iconStyle}  />
              </span>
        }
      </span>
    );


    if (link) {
      return (
        <a {...combineProps}>
          {child}
        </a>
      );
    }

    return (
        <button {...combineProps}>
          {child}
        </button>
    );
}

const styles = {
  base: {
      padding: 10,
      fontSize: 18,
      minWidth: 150,
      cursor: 'pointer',
      position: 'relative',
      textDecoration: 'none',
      border: '1px solid #333'
  },
  primary: {
    color: '#333',
    backgroundColor: '#FFF'
  },
  secondary: {
    color: '#FFF',
    backgroundColor: '#333'
  },
  rounded: (size) => {
    return {
      borderRadius: size || 60
    };
  },
  block: {
    borderRadius: 0
  },
  iconContainer: {
    margin: '0 0 0 10px'
  }
};


Button.defaultProps = {
  style: {},
  rounded: false,
  type: 'primary',
  iconPosition: 'right'
};


Button.propTypes = {
  /**
   * Specifies if the button can and should act as
   * an anchor. If set to true, then the component
   * will have the same properties of the <i><</i>a<i></i>> tag
   */
  link: PropTypes.bool,
  icon: PropTypes.string,

  /**
   * Content to be displayed inside the button.
   */
  text: PropTypes.string,

  /**
   * This prop can be used to override the default styling of this button component.
   */
  style: PropTypes.object,

  /**
   * Callback function fired when the button is clicked.
   * <br/>
   * <pre>
   *   <code>
   *      function(event: object): void
   *   </code>
   *   <i>event</i>: Click event targeting the button.
   * </pre>
   */
  onClick: PropTypes.func,

  /**
   * Override default styling of the icon.
   */
  iconStyle: PropTypes.object,

  /**
   * Override the default style of the wrapper around the icon.
   */
  iconWrapperStyle: PropTypes.object,

  /**
   * Position of the icon, either <i>"right"</i> or <i>"left"</i>
   */
  iconPosition: PropTypes.oneOf(['left', 'right']),

  /**
   * Can be a "number" or "boolean". If boolean and true,
   * the border radius is set to 60.
   * If a number is specified,
   * then the border radius will use that number.
   */
  rounded: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.number
  ]),

  /**
   * Element to be displayed inside the button.
   * If text is defined, then the button will use the value of text
   * in the button over children. But, otherwise, children will be
   * displayed.
   */
  children: PropTypes.node,
  
  /**
   * The type of the button.
   * Can be one of "primary", "secondary"
   */
  type: PropTypes.oneOf(['primary', 'secondary'])
};

export default Radium(Button);
