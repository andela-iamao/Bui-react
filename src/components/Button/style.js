export default {
    base: {
        padding: 10,
        fontSize: 18,
        minWidth: 150,
        cursor: 'pointer',
        position: 'relative',
        textDecoration: 'none',
        border: '1px solid #333'
    },
    primary: {
      color: '#333',
      backgroundColor: '#FFF'
    },
    secondary: {
      color: '#FFF',
      backgroundColor: '#333'
    },
    rounded: (size) => {
      return {
        borderRadius: size || 60
      };
    },
    block: {
      borderRadius: 0
    },
    iconContainer: {
      margin: '0 0 0 10px'
    }
};