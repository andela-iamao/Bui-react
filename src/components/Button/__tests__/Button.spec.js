import React from 'react';
import renderer from 'react-test-renderer';

import { shallow, render } from 'enzyme';

import Button from '../';
import BuiTheme from '../../BuiTheme';


describe('Button', () => {

    describe('Snapshot Test', () => {

        test('renders display with a single child component', () => {
            const tree = renderer.create(<Button>Hello!</Button>).toJSON();
            expect(tree).toMatchSnapshot();
        });

        test('renders a round button component', () => {
            const tree = renderer.create(<Button rounded>Hello!</Button>).toJSON();
            expect(tree).toMatchSnapshot();
        });

        test('renders with the BuiTheme default theme', () => {
            const tree = renderer.create(
                <BuiTheme>
                    <Button>Hello!</Button>
                </BuiTheme>
            ).toJSON();
            expect(tree).toMatchSnapshot();
        });

        test('renders a round button component with default theme', () => {
            const tree = renderer.create(
                <BuiTheme>
                    <Button rounded>Hello!</Button>
                </BuiTheme>
            ).toJSON();
            expect(tree).toMatchSnapshot();
        });
    
    });
    

    describe('Unit Test', () => {

        test('Button should display a text "Hello!"', () => {
            const wrapper = shallow(
                <Button>Hello!</Button>
            );
            expect(wrapper.length).toEqual(1);
            expect(wrapper.text()).toEqual('Hello!');
        });

        test('Button should display text props if no children props', () => {
            const wrapper = shallow(
                <Button text="Hello!"/>
            );
            expect(wrapper.length).toEqual(1);
            expect(wrapper.text()).toEqual('Hello!');
        });

        test('Button should take in onClick props and call it on click event', () => {
            const mock = jest.fn();
            const onClick = new mock();
            
            const wrapper = shallow(
                <Button onClick={() => onClick}>
                    Click me!
                </Button>
            )
            wrapper.find('button').simulate('click');

            expect(mock.mock.calls.length).toBe(1);
        });

        test('Button should display an icon', () => {
            const wrapper = render(
                <Button icon="eye">Hello!</Button>
            );

            expect(wrapper.length).toEqual(1);
            expect(wrapper.text().search('Hello!')).toEqual(0);
            expect(wrapper.children().children().children()[0].name).toEqual('svg');
        });
    });
    
});
