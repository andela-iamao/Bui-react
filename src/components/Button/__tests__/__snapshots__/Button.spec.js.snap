// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot Test renders a round button component 1`] = `
<button
  data-radium={true}
  style={
    Object {
      "backgroundColor": "#FFF",
      "border": "1px solid #333",
      "color": "#333",
      "cursor": "pointer",
      "fontSize": 18,
      "minWidth": 150,
      "padding": 10,
      "position": "relative",
      "textDecoration": "none",
    }
  }
>
  <span
    data-radium={true}
  >
    Hello!
  </span>
</button>
`;

exports[`Button Snapshot Test renders a round button component with default theme 1`] = `
<button
  data-radium={true}
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <span
    data-radium={true}
  >
    Hello!
  </span>
</button>
`;

exports[`Button Snapshot Test renders button with a custom theme 1`] = `
<button
  data-radium={true}
  style={
    Object {
      "borderRadius": 10,
      "color": "red",
      "width": 400,
    }
  }
>
  <span
    data-radium={true}
  >
    Hello!
  </span>
</button>
`;

exports[`Button Snapshot Test renders display with a single child component 1`] = `
<button
  data-radium={true}
  style={
    Object {
      "backgroundColor": "#FFF",
      "border": "1px solid #333",
      "borderRadius": 0,
      "color": "#333",
      "cursor": "pointer",
      "fontSize": 18,
      "minWidth": 150,
      "padding": 10,
      "position": "relative",
      "textDecoration": "none",
    }
  }
>
  <span
    data-radium={true}
  >
    Hello!
  </span>
</button>
`;

exports[`Button Snapshot Test renders with the BuiTheme default theme 1`] = `
<button
  data-radium={true}
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <span
    data-radium={true}
  >
    Hello!
  </span>
</button>
`;
